modules/rgbd/src/tsdf_functions.cpp

```diff
@@ -373,5 +373,224 @@ void integrateVolumeUnit(
+        if (v.idx[0] == -2147483647)
+        {
+            v.idx = idx;
+            v.nextVolumeRow = getNextVolume(hash, num, i, start);
+            indexes.push_back(indx);
+            indexesGPU.push_back(idx);
+            return;
+        }
+        i = v.nextVolumeRow;
+    }
+}
+
+void VolumesTable::update(Vec3i indx, int row)
+{
+    Vec4i idx(indx[0], indx[1], indx[2], 0);
+    int hash = int(calc_hash(idx) % hash_divisor);
+    int num = 1;
+    int start = hash * num * list_size;
+    int i = start;
+
+    while (i != -1)
+    {
+        Volume_NODE& v = volumes.at<Volume_NODE>(i, 0);
+        if (v.idx == idx)
+        {
+            v.row = row;
+            return;
+        }
+        //find nan cheking for int or Vec3i
+        //if (isNaN(Point3i(v.idx)))
+        if (v.idx[0] == -2147483647)
+        {
+            v.idx = idx;
+            v.row = row;
+            v.nextVolumeRow = getNextVolume(hash, num, i, start);
+            indexes.push_back(indx);
+            indexesGPU.push_back(idx);
+            return;
+        }
+        i = v.nextVolumeRow;
+    }
+}
+
+void VolumesTable::update(Vec3i indx, int isActive, int lastVisibleIndex)
+{
+    Vec4i idx(indx[0], indx[1], indx[2], 0);
+    int hash = int(calc_hash(idx) % hash_divisor);
+    int num = 1;
+    int start = hash * num * list_size;
+    int i = start;
+
+    while (i != -1)
+    {
+        Volume_NODE& v = volumes.at<Volume_NODE>(i, 0);
+        if (v.idx == idx)
+        {
+            v.isActive = isActive;
+            v.lastVisibleIndex = lastVisibleIndex;
+            return;
+        }
+        //find nan cheking for int or Vec3i
+        //if (isNaN(Point3i(v.idx)))
+        if (v.idx[0] == -2147483647)
+        {
+            v.idx = idx;
+            v.nextVolumeRow = getNextVolume(hash, num, i, start);
+            v.isActive = isActive;
+            v.lastVisibleIndex = lastVisibleIndex;
+            indexes.push_back(indx);
+            indexesGPU.push_back(idx);
+            return;
+        }
+        i = v.nextVolumeRow;
+    }
+}
+
+void VolumesTable::updateActive(Vec3i indx, int isActive)
+{
+    Vec4i idx(indx[0], indx[1], indx[2], 0);
+    int hash = int(calc_hash(idx) % hash_divisor);
+    int num = 1;
+    int start = hash * num * list_size;
+    int i = start;
+
+    while (i != -1)
+    {
+        Volume_NODE& v = volumes.at<Volume_NODE>(i, 0);
+        if (v.idx == idx)
+        {
+            v.isActive = isActive;
+            return;
+        }
+        //find nan cheking for int or Vec3i
+        //if (isNaN(Point3i(v.idx)))
+        if (v.idx[0] == -2147483647)
+        {
+            return;
+        }
+        i = v.nextVolumeRow;
+    }
+}
+
+bool VolumesTable::getActive(Vec3i indx) const
+{
+    Vec4i idx(indx[0], indx[1], indx[2], 0);
+    int hash = int(calc_hash(idx) % hash_divisor);
+    int num = 1;
+    int i = hash * num * list_size;
+    while (i != -1)
+    {
+        Volume_NODE v = volumes.at<Volume_NODE>(i, 0);
```

Comment:
This thing and similar ones fail to compile on my machine.
This works: `Volume_NODE& v = *volumes.ptr<Volume_NODE>(i);`

Reply:
done